ethdb/privateapi/ethbackend.go

```diff
@@ -271,70 +279,157 @@ func (s *EthBackendServer) EngineExecutePayloadV1(ctx context.Context, req *type
 
 // EngineGetPayloadV1, retrieves previously assembled payload (Validators only)
 func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.EngineGetPayloadRequest) (*types2.ExecutionPayload, error) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.condPauseAssemble.L.Lock()
+	defer s.condPauseAssemble.L.Unlock()
+	// Wait some time in case validition process has not started
+
+	if !s.proposing {
+		return nil, fmt.Errorf("execution layer not running as a proposer. enable --proposer flag on startup")
+	}
 
 	if s.config.TerminalTotalDifficulty == nil {
 		return nil, fmt.Errorf("not a proof-of-stake chain")
 	}
-
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	payload, ok := s.pendingPayloads[req.PayloadId]
-	if ok {
-		return &payload, nil
+	if !ok {
+		return nil, fmt.Errorf("unknown payload")
+	}
+	if payload.BlockNumber == 0 {
+		s.mu.Unlock()
+		// Wait for payloads assembling thread to finish
+		s.condPauseAssemble.Wait()
```

Comment:
You need two separate `sync.Cond` because you have two distinct wait/broadcast scenarios:
1. Waiting for previous payload assembly to finish
2. Waiting for new `EngineForkChoiceUpdatedV1` requests (with payload) to resume payload assembly goroutine 

So you need separate `sync.Cond` instances to avoid one spuriously awaking the other.